### PR TITLE
CMake support

### PR DESCRIPTION
**Pull request by desertkun**
This PR adds a toolchain `cmake/Toolchain-zcc.cmake` which makes the compiler work with CMake "out of the box".

CMakeLists.txt:
```
project(test2 C)
add_executable(test2bin main.c)
```

main.c:
```
#include <stdio.h>
int main() {
	printf("Hello, world!\n");
}
```

Then, 
```
mkdir build
cd build
cmake -DZCCTARGET=test -DCMAKE_TOOLCHAIN_FILE=/usr/local/share/z88dk/cmake/Toolchain-zcc.cmake ..
make

Scanning dependencies of target test2bin
[ 50%] Building C object CMakeFiles/test2bin.dir/main.c.o
[100%] Linking C executable test2bin.bin
[100%] Built target test2bin
```

Finally,
```
z88dk-ticks test2bin.bin
Hello, world!

Ticks: 21121
```

I have created a [wiki page](https://github.com/z88dk/z88dk/wiki/CMake) to explain this to users, after this PR merges I am going to do fill it up, currently it is empty.

### CODE REVIEW

**Comment by suborb** [comment]
Can the cmake directory be moved into support/cmake ?

Otherwise (if you just glance) suggests that cmake can be used to build z88dk

**Comment by desertkun** [comment]
That would have been pretty cool actually, but I'll fix it.

**Comment by desertkun** [comment]
Fixed. I now wonder if I can try and refactor z88dk itself onto cmake 🤔 Managing makefiles is a pain and CMake can generate VS project as well, so one less project to manage.

**Comment by desertkun** [comment]
Here's the example repo (3 files really), tested on ubuntu and mac https://github.com/desertkun/testing222.git

Would appreciate if anyone can test this out, just run "build.sh", before I can merge

**Comment by pauloscustodio** [comment]
I've got a failed test on Windows/msys2, but not being a CMake user myself, don't know if it anything I did wrong:
```
$ cmake -DZCCTARGET=test -DCMAKE_TOOLCHAIN_FILE=/tmp/z88dk/support/cmake/Toolchain-zcc.cmake ..
zcc - Frontend for the z88dk Cross-C Compiler
See https://github.com/z88dk/z88dk/wiki/CMake for more information
zcc - Frontend for the z88dk Cross-C Compiler
See https://github.com/z88dk/z88dk/wiki/CMake for more information
-- The C compiler identification is unknown
-- Detecting C compiler ABI info
zcc - Frontend for the z88dk Cross-C Compiler
See https://github.com/z88dk/z88dk/wiki/CMake for more information
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/paulo/git/z88dk/bin/zcc.exe - skipped
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/testing222/build

$ make
Scanning dependencies of target test2bin
[ 50%] Building C object CMakeFiles/test2bin.dir/main.c.o
A config file must be specified with +file as the first argument

zcc - Frontend for the z88dk Cross-C Compiler - vMSC build-Sat 07/23/2022

Usage: C:\msys64\home\paulo\git\z88dk\bin\zcc.exe +[target] {options} {files}
   -v -verbose                  Output all commands that are run (-vn suppresses)
   -h -help                     Display this text
      -o                        Set the basename for linker output files
      -specs                    Print out compiler specs

CPU Targetting:
      -m8080                    Generate output for the i8080
      -m8085                    Generate output for the i8085
      -mz80                     Generate output for the z80
      -mz80n                    Generate output for the z80n
      -mz180                    Generate output for the z180
      -mr2ka                    Generate output for the Rabbit 2000
      -mr3k                     Generate output for the Rabbit 3000
      -mgbz80                   Generate output for the Gameboy Z80

Target options:
      -subtype                  Set the target subtype
      -clib                     Set the target clib type
      -crt0                     Override the crt0 assembler file to use
      -startuplib               Override STARTUPLIB - compiler base support routines
      --no-crt                  Link without crt0 file
      -startup                  Set the startup type
      -zorg                     Set the origin (only certain targets)
      -nostdlib                 If set ignore INCPATH, STARTUPLIB
      -pragma-redirect          Redirect a function
      -pragma-define            Define the option in zcc_opt.def
      -pragma-output            Define the option in zcc_opt.def (same as above)
      -pragma-export            Define the option in zcc_opt.def and export as public
      -pragma-need              NEED the option in zcc_opt.def
      -pragma-bytes             Dump a string of bytes zcc_opt.def
      -pragma-include           Process include file containing pragmas

Lifecycle options:
      -m4                       Stop after processing m4 files
   -E --preprocess-only         Stop after preprocessing files
   -c --compile-only            Stop after compiling .c .s .asm files to .o files
   -a --assemble-only           Stop after compiling .c .s files to .asm files
   -S --assemble-only           Stop after compiling .c .s files to .asm files
   -x                           Make a library out of source files
      -create-app               Run appmake on the resulting binary to create emulator usable file

M4 options:
      -Cm                       Add an option to m4
      -copy-back-after-m4       Copy files back after processing with m4

Preprocessor options:
      -Cp                       Add an option to the preprocessor
      -D                        Define a preprocessor option
      -U                        Undefine a preprocessor option
      -I                        Add an include directory for the preprocessor
      -iquote                   Add a quoted include path for the preprocessor
      -isystem                  Add a system include path for the preprocessor

Compiler (all) options:
      -compiler                 Set the compiler type from the command line (sccz80,sdcc)
      --c-code-in-asm           Add C code to .asm files
      --opt-code-speed          Optimize for code speed
      -debug                    Enable debugging support

Compiler (sccz80) options:
      -Cc                       Add an option to sccz80
      -set-r2l-by-default       (sccz80) Use r2l calling convention by default
      -O                        Set the peephole optimiser setting for copt

Compiler (sdcc) options:
      -Cs                       Add an option to sdcc
      --opt-code-size           Optimize for code size (sdcc only)
      -SO                       Set the peephole optimiser setting for sdcc-peephole
      --fsigned-char            Use signed chars by default

Compiler (clang/llvm) options:
      -Cg                       Add an option to clang
      -clang
Options:

  Stop after translating .c files to llvm ir
      -llvm                     Stop after llvm-cbe generates new .cbe.c files
      -Co                       Add an option to llvm-opt
      -Cv                       Add an option to llvm-cbe
      -zopt                     Enable llvm-optimizer (clang only)

Assembler options:
      -Ca                       Add an option to the assembler
      -z80-verb                 Make the assembler more verbose

Linker options:
      -Cl                       Add an option to the linker
      -L                        Add a library search path
      -l                        Add a library
      -bn                       Set the output file for the linker stage
      -reloc-info               Generate binary file relocation information
   -m -gen-map-file             Generate an output map of the final executable
   -s -gen-symbol-file          Generate a symbol map of the final executable
      --list                    Generate list files

Appmake options:
      -Cz                       Add an option to appmake

Misc options:
      -g                        Generate a global defc file of the final executable (-g -gp -gpf:filename)
      -alias                    Define a command line alias
      --lstcwd                  Paths in .lst files are relative to the current working dir
      -custom-copt-rules        Custom user copt rules
      -no-cleanup               Don't cleanup temporary files

Argument Aliases:

[100%] Linking C executable test2bin.bin
error: file open: CMakeFiles/test2bin.dir/main.c.o
make[2]: *** [CMakeFiles/test2bin.dir/build.make:96: test2bin.bin] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/test2bin.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

**Comment by desertkun** [comment]
You're referencing the "source" toolchain from repo and not installed one into the z88dk installed folder. It seems, though, that it still should work. Thanks.

**Comment by desertkun** [comment]
@pauloscustodio did you rebuild z88dk from source? My latest merge relaxed requirement of having +conf as being first argument.

**Comment by pauloscustodio** [comment]
My z88dk may have been out-of-date. Let me try again.

**Comment by pauloscustodio** [comment]
First comment: should be -I../../ext/uthash/src in zcc Makefile

**Comment by desertkun** [comment]
> First comment: should be -I../../ext/uthash/src in zcc Makefile

Not sure I follow but that PR went in and I checked there is a `z88dk/ext/uthash/include` folder although it is just a symlink to src

**Comment by pauloscustodio** [comment]
Yes, but symlinks do not work on Windows... so we need to point to src

**Comment by desertkun** [comment]
Whoops

**Comment by desertkun** [comment]
@pauloscustodio Yeah, fixed that, thanks. Wonder how it builds build-on-windows-msbuild jobs.

**Comment by pauloscustodio** [comment]
> Wonder how it builds build-on-windows-msbuild jobs.
The compiler flags are defined in the Visual Studio project, the Makefiles are not used.

**Comment by desertkun** [comment]
> The compiler flags are defined in the Visual Studio project, the Makefiles are not used.

Yeah, but [my fix](https://github.com/z88dk/z88dk/commit/fde969b352770a959caf2237d2b5a2effb34c962#diff-a787f8e4e5e0a7280c90cc7b1308b7393ffd73f5b70c17ab9cd5e643ca6753c1) changed the project as well, prior it was referring to include. Weird.

**Comment by pauloscustodio** [comment]
All good in Windows/msys2

**Comment by desertkun** [comment]
Thanks.

**Comment by pauloscustodio** [comment]
> @pauloscustodio Yeah, fixed that, thanks. Wonder how it builds build-on-windows-msbuild jobs.

The build-on-windows-msbuild jobs build the Release configuration, the fixes you made where on the Debug configuration.

I will extend the build-on-windows-msbuild jobs  to build also the Debug configuration.
